### PR TITLE
[JIT] simplify logic for merging types during profiling

### LINKED ISSUE #67751: Remove logic for constructing symbolic shapes in Profiler
## 🚀 Feature

Simplify Profiler by removing two pieces of logic in `profiling_record.cpp` responsible for grouping profiled tensor types by `frame_id` (i.e. an unique identifier of a particular instance of a function) and for merging groups into final symbolic shapes.

## Motivation

Profiler is an important component of ProfilingExecutor which is our default runtime to execute TorchScript'ed models.
Profiler is used in a number of scenarios. 
* For TS'ed models, the profiling information is used for fusion element-wise operations on both CPU and GPU
* Models running on a lazy device will also be utilizing PE and Profiler to collect profiling information for fusers.
* Profiler can be used to collect shapes, strides ahead of time and to also trace through the control flow in the original graphs.

**This particular item will allow us to simplify the implementation of Profiler which should make it easier to add new features and service in the future. It may also result in moderate gains, as the merging step will be combined with the profiling step.**

## Pitch

1. Remove `profiled_types_per_frame_` from `profiling_record.h`
2. In https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_record.cpp#L174-L204 use "profiled_type" attribute on `prim::profile` node to store and merge profiling information for a given `Value*` regardless of which invocation of a function (i.e. which `frame_id`) it came from
  a. An example of setting `ty_` : https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_record.cpp#L172
  b. An example of using `ty`: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L284
3. Remove this code : https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_record.cpp#L358-L416 since we should already have merged profiling information for each value in `profiled_type` attribute of each prim::profile node.
4. Remove any auxiliary functions and/or structures used by either 2. or 3. 
5. Run CI to identify tests that were testing the removed feature.
6. Remove the tests.

## Additional context

1. Learning how to build PyTorch and run tests (1-3 days)
2. Reading `profiling_record.cpp`/`profiling_record.h` (1-2 days)
3. Cleaning up `profiling_record.cpp` (1-2 days)
4. Running the tests (1 day)
5. Inspecting failures (1-3 days)

Before working on this task, it's recommended to acquire some basic understanding of PE flow. This can be done by writing a simple model (2-3 ops) or using one of our tests in `test_jit.py` or `test_jit_fuser_te.py` and running it multiple times (1st time for profiling, 2nd time for optimization) and inspecting the output of `export PYTORCH_JIT_LOG_LEVEL=>>profiling_graph_executor_impl`

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69096

Instead of storing profiling data in a map and then merginging at
the end, perform merging directly during profiling.

Differential Revision: [D32772626](https://our.internmc.facebook.com/intern/diff/D32772626)

### CODE REVIEW

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 1]
```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
...
+
     if (v.isTensor()) {
-      std::lock_guard<std::mutex> lock(this->mutex_);
-      auto& profiled_types = profiled_types_per_frame_[frame_id];
       auto& t = v.toTensor();
-      if (t.defined()) {
-        auto pttp = tensorTypeInCurrentExecutionContext(t);
+      auto new_tensor_type =
+          (t.defined() ? tensorTypeInCurrentExecutionContext(t)
+                       : TensorType::get()->withUndefined());
```
Comment (swolchok) [inline torch/csrc/jit/runtime/profiling_record.cpp:161]: doesn't tensorTypeInCurrentExecutionContext already do this?

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 2]
```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
...
+          "In run ",
+          frame_id,
+          " annotating %",
+          pno->debugName(),
+          " with ",
+          *new_tensor_type);
+
+      if (pn->hasRun()) {
+        auto existing_tensor_type =
+            pn->ty(attr::profiled_type)->expect<TensorType>();
```
Comment (swolchok) [inline torch/csrc/jit/runtime/profiling_record.cpp:172]: You don't need the shared_ptr, so use expectRef and save a reference count bump
`const auto& existing_tensor_type = pn->ty(attr::profiled_type)->expectRef<TensorType>();`

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 3]
```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
...
-          GRAPH_DEBUG("Existing type for %", pno->debugName(), " ", *type);
-          pttp = type->merge(*pttp);
-          GRAPH_DEBUG("Result for %", pno->debugName(), " ", *pttp);
-          profiled_types[pno] = pttp;
-        }
+            ": ",
+            *existing_tensor_type);
+        auto merged_type = new_tensor_type->merge(*existing_tensor_type);
+        GRAPH_DEBUG("Merged type for %", pno->debugName(), ": ", *merged_type);
+        pn->ty_(attr::profiled_type, merged_type);
```
Comment (swolchok) [inline torch/csrc/jit/runtime/profiling_record.cpp:180]: save a refcount bump here. would be slightly faster still to inline the call when the GRAPH_DEBUG won't fire, but we probably don't need to go there until profiling data says we do.
`         pn->ty_(attr::profiled_type, std::move(merged_type));`

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 4]
```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
...
       } else {
-        profiled_types[pno] = TensorType::get()->withUndefined();
+        pn->setHasRun(true);
+        pn->ty_(attr::profiled_type, new_tensor_type);
```
Comment (swolchok) [inline torch/csrc/jit/runtime/profiling_record.cpp:183]: same, missing move

`pn->ty(attr::profiled_type, std::move(new_tensor_type));`

File: torch/csrc/jit/runtime/profiling_record.h [hunk 5]
```diff
@@ -184,15 +184,11 @@ struct ProfilingRecord {
   std::shared_ptr<Graph> profiled_graph_;
   mutable std::mutex mutex_;
   size_t profiling_count_;
-  // the key is a frame id
-  // the value is a mapping from a Value in a graph
-  // to a profiled TensorType
-  std::map<int64_t, std::map<Value*, TensorTypePtr>> profiled_types_per_frame_;
```
Comment (swolchok) [inline torch/csrc/jit/runtime/profiling_record.h:190]: cool, no need to use std::unordered_map or F14 if the map is just gone!

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 6]
```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
...
         GRAPH_DEBUG(
-            "In run ",
-            frame_id,
-            " annotating %",
+            "Existing type for %",
             pno->debugName(),
-            " with ",
-            *pttp);
-        if (profiled_types.count(pno) == 0) {
-          profiled_types.insert({pno, pttp});
```
Comment (swolchok) [inline torch/csrc/jit/runtime/profiling_record.cpp:166]: this had a missing move, so if we ever need it back we should put the move in

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32772626).

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 7]
```diff
@@ -144,35 +144,42 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
...
             frame_id,
             " annotating %",
             pno->debugName(),
             " with ",
-            *pttp);
-        if (profiled_types.count(pno) == 0) {
-          profiled_types.insert({pno, pttp});
+            *new_tensor_type);
+
+        if (pn->hasRun()) {
```
Comment (Krovatkin) [inline torch/csrc/jit/runtime/profiling_record.cpp:167]: could we use `hasAttr(attr::profiled_type)` on `pn`?

File: torch/csrc/jit/runtime/profiling_record.h [hunk 8]
```diff
@@ -184,15 +184,11 @@ struct ProfilingRecord {
...
 
   // A thin wrapper around `partitionSetByDimension` to ensure
   // `new_sizes` and `sym_shapes` are of the same rank
 
-  c10::SymbolicShape mergeSymbolicShapes(
+  static c10::SymbolicShape mergeSymbolicShapes(
```
Comment (Krovatkin) [inline torch/csrc/jit/runtime/profiling_record.h:191]: I was hoping we can nuke this helper? Are there any uses?

**Review by Krovatkin — COMMENTED** [review]
This looks amazing! A few quick questions!

File: torch/csrc/jit/runtime/profiling_record.h (hunk 8 above)
Comment (davidberard98) [reply torch/csrc/jit/runtime/profiling_record.h:191]: whoops, missed that - I think you're right

File: torch/csrc/jit/runtime/profiling_record.cpp (hunk 7 above)
Comment (davidberard98) [reply torch/csrc/jit/runtime/profiling_record.cpp:142]: @Krovatkin there's a few places that expect `attr::profiled_type` to be set; I tried removing the initialization (i.e. `pn->ty_(attr::profiled_type, TensorType::get());`) but a bunch of tests fail, e.g.

```
ERROR: test_bias_as_arg (jit.test_autodiff_subgraph_slicing.TestAutodiffSubgraphSlicing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/dberard/pytorch/test/jit/test_autodiff_subgraph_slicing.py", line 135, in test_bias_as_arg
    check_against_reference(self, scripted, method1, lambda x: x, (x, weight, bias), check_types=False)
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.9/site-packages/torch/testing/_internal/common_jit.py", line 128, in check_against_reference
    grads_test = torch.autograd.grad(
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.9/site-packages/torch/autograd/__init__.py", line 275, in grad
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: required keyword attribute 'profiled_type' is undefined

----------------------------------------------------------------------
Ran 26 tests in 2.127s

FAILED (errors=1)
```

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32772626).

File: torch/csrc/jit/runtime/profiling_record.cpp (hunk 7 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_record.cpp:142]: @davidberard98 could you please elaborate why switching 
```
if (pn->hasRun() {
// ... merge types
} else {
 pn->ty_(attr::profiled_type, new_incoming_type);
}
```

to 

```
if (pn->hasAttribute(attr::profiled_type)) {
// ... merge types
} else {
 pn->ty_(attr::profiled_type, new_incoming_type);
 // next time we call `hasAttribute` we merge types
} 
```

Comment (davidberard98) [reply torch/csrc/jit/runtime/profiling_record.cpp:142]: @Krovatkin 
Currently, we initialize `pn->ty_(attr::profiled_type, TensorType::get())` before the lambda gets called (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_record.cpp#L145). So if we simply replaced `hasRun()` with `hasAttribute()`, then `pn->hasAttribute(attr::profiled_type)` would always evaluate to true.

If we remove the initialization, then we get the `required keyword attribute 'profiled_type' is undefined` error that I mentioned.

**Review by Krovatkin — APPROVED** [review]
:shipit:

**Comment by Krovatkin** [comment]
@swolchok do you want to do another pass or can we merge this?

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32772626).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32772626).

**Comment by swolchok** [comment]
> @swolchok do you want to do another pass or can we merge this?

no complaints from me!

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32772626).

**Pull request closed without merging**